Cell Dissemination (Partial messages)

dknopik opened the pull request:
## Issue Addressed

- https://github.com/ethereum/consensus-specs/pull/4558
- https://eips.ethereum.org/EIPS/eip-8136

## Important locations

[‎consensus/types/src/data/partial_data_column_sidecar.rs](https://github.com/sigp/lighthouse/pull/8314/changes#diff-11a35759483d97e9a74bb936984d0fc554e9531482a79bf74119323ddd3d4d2f):

`PartialDataColumnSidecar`: the actual message. 
`PartialDataColumnSidecarRef`: the actual message, but using references: This allows partial.rs (see below) to create lightweight structs containing only some parts of the canonical partial data column. It encodes to the same messages as `PartialDataColumnSidecar`.
`PartialDataColumnHeader` : an optional header in the message, containing the commitments, inclusion proof and signed block header. This is sent on every first partial sent between two nodes. If it is NOT sent, Lighthouse will add it to the message if it was not 
`PartialDataColumn`: the sidecar, plus some information that we derive: the column index from the subnet, and the block hash from the group id.
`PartialDataColumnPartsMetadata`: The metadata used for the partial messages - see spec. Right now, we always request missing cells, so the request bitmap is effectively always all ones.

[beacon_node/lighthouse_network/src/types/partial.rs‎](https://github.com/sigp/lighthouse/pull/8314/changes#diff-8cf8bb42bfb4bcfb881a84ca0703cf0d3dd5b86111f3e178ade12dfcf9494a3b): The main interaction with libp2p

[‎beacon_node/beacon_chain/src/partial_data_column_assembler.rs](https://github.com/sigp/lighthouse/pull/8314/changes#diff-7bc36a775d1e20fdac5068278faa31e243b1b16469f8d9823a4e5c5c392dc591): Where the partials end up. As they complete they are moved to the overflow_lru_cache. On every update partials are republished, using the `PartialMergeResult` data.

## Ugly bits

- There are a lot of expects around the sidecar and the metadata as we can assert bounds. I think some of them can be removed if we add some convenience functions to `ssz_types` and/or `ethereum_ssz`.
- `PartialDataColumnHeader` is also bleeding into code paths not *directly* related to partial messages:  as `getBlobsV3` and `getBlobsV2` (if V3 is not available) are triggered from partial messages if they arrive before the block (which is not unlikely), `fetch_and_process_blobs_v2_or_v3` takes a header as argument instead of a full block, which is sufficient and can be derived from a block, but awkward if called from contexts unrelated to partial messages. In conjunction with ePBS implementation efforts, this should probably renamed, a common container created, or put behind a trait or something like that.

## Left to do

- Gloas impl (to be done in a future PR)
- [x] Unit tests
- [x] Wait for https://github.com/sigp/lighthouse/pull/8731
- [x] Wait for https://github.com/libp2p/rust-libp2p/pull/6275

dknopik committed:
Implement Cell Dissemination

dknopik committed:
Merge branch 'unstable' into partial-columns

dknopik committed:
fix compilation

dknopik committed:
fix some lints

dknopik committed:
impl get_blobs_v2 fallback

dknopik committed:
update gossipsub

dknopik committed:
remove redundant log

dknopik committed:
try to upgrade column using block immediately

dknopik committed:
publish data after block availability

dknopik committed:
fix remaining lints and release test compilation

dknopik committed:
implement caching of partial messages

dknopik committed:
update gossipsub

dknopik committed:
add CLI flags for disable request and disable support

dknopik committed:
Merge branch 'unstable' into partial-columns

dknopik committed:
update CLI docs

dknopik committed:
do not error for column on failed observation check

dknopik committed:
Fix get blobs v2 tests

dknopik committed:
Reintroduce sanity check for get blobs v2

dknopik committed:
Only count actually received blobs

dknopik committed:
Log full vs partial

dknopik committed:
fix and improve logging

dknopik committed:
update gossipsub

dknopik committed:
Merge branch 'refs/heads/unstable' into partial-columns

dknopik committed:
look for block in data availability checker

dknopik committed:
re-enable SSE

dknopik committed:
Properly handly observing in new_for_block_publishing

dknopik committed:
Fix cache and better logging

dknopik committed:
Fix invalid `UnexpectedDataColumn` error

dknopik committed:
remove custom cache in favour of reprocessing queue

dknopik committed:
Merge branch 'unstable' into partial-columns

dknopik committed:
publish partials on http API block

dknopik committed:
experimentally full eager publish for API blocks

dknopik committed:
experimental libp2p changes

dknopik committed:
joaos libp2p changes

dknopik committed:
Merge branch 'unstable' into partial-columns

dknopik committed:
fix test

dknopik committed:
Merge branch 'unstable' into partial-columns

dknopik committed:
fix function path

dknopik committed:
update libp2p

dknopik committed:
update libp2p again

dknopik committed:
do not eagerly publish all columns on proposal

dknopik committed:
once again update libp2p

dknopik committed:
merge from unstable

dknopik committed:
fix import

dknopik committed:
update for libp2p fix

dknopik committed:
Another update for libp2p to fix received msgs

dknopik committed:
Fix inverted condition

dknopik committed:
fix kzg verification

dknopik committed:
actually dequeue partials

dknopik committed:
update libp2p

dknopik committed:
Merge unstable

dknopik committed:
Cargo fmt

pawanjay176 committed:
Refactor

dknopik committed:
Merge from unstable

dknopik committed:
add getBlobsV3 metrics

dknopik committed:
add cell usefulness metric definitions

dknopik committed:
update libp2p

dknopik committed:
add PartialDataColumnHeader

dknopik committed:
remove VerifiablePartialDataColumn

dknopik committed:
add component for tracking sent headers

dknopik committed:
remove eager sending of cells

dknopik committed:
Use separate network flow for partial messages

dknopik committed:
make early partial handling slightly less awkward

dknopik committed:
merge pawans refactor

dknopik committed:
Add validation and caching of partial header

dknopik committed:
Arc the cells in a partial data column sidecar

dknopik committed:
Rename "DanglingPartialDataColumn"

dknopik committed:
Arc less

dknopik committed:
Remove GossipVerifiedPartialDataColumn

dknopik committed:
Merge from unstable

dknopik committed:
update libp2p

dknopik committed:
Fix obvious mistakes

dknopik committed:
Remove outdated todo

dknopik committed:
Fix fetch blobs tests

dknopik committed:
Remove unnecessary explicit init

dknopik committed:
Undo buggy changes in block publish flow

dknopik committed:
Implement getting blobs after header arrival

dknopik committed:
fix SSE

dknopik committed:
Implement 3 of 4 missing metrics

dknopik committed:
add version byte prefix to group id

dknopik committed:
update libp2p

dknopik committed:
merge from unstable

dknopik committed:
add logging on send

dknopik committed:
Improve logging

dknopik committed:
avoid completing already completed columsn

dknopik committed:
fix metric name

dknopik committed:
fix: always publish partials after attempted getblobsv2/v3 and avoid useless republishs

dknopik committed:
fmt

dknopik committed:
add moar logs

dknopik committed:
fix compilation

dknopik committed:
update libp2p

dknopik committed:
clean up

dknopik committed:
Cherry pick c3fb9eb0 for devnet retention fix

dknopik committed:
invert CLI param

dknopik committed:
Do maintenance and removal for partial assembler

dknopik committed:
merge from unstable

dknopik committed:
implement spec change: request bitmap

dknopik committed:
Remove column from assembly when complete

This enables
- last metric
- correct handling of full column

dknopik committed:
merge from unstable

dknopik committed:
only send partial if enabled

dknopik committed:
log silent if no peers support partial

no special case for duplicates

dknopik committed:
update libp2p

dknopik committed:
merge from unstable

dknopik committed:
update libp2p

dknopik committed:
implement optional metadata function

dknopik committed:
implement reporting of failures to libp2p

dknopik committed:
merge from unstable

dknopik committed:
fix lookup on unknown parent

dknopik committed:
update libp2p

dknopik committed:
merge from unstable

dknopik committed:
fix inverted logic for sending

dknopik committed:
merge from unstable

dknopik committed:
ensure atomic and guaranteed setting of has_local_blobs

dknopik committed:
Add metric for headerless partial verification

dknopik committed:
merge from unstable and update libp2p

dknopik committed:
move send log to more sensible location

dknopik committed:
update libp2p

dknopik committed:
merge from unstable

dknopik committed:
optional partial assembler

dknopik committed:
refactor: encode without cloning, no cell level arc, better validation flow

dknopik committed:
improve logging around getBlobsV3

dknopik committed:
update libp2p

dknopik committed:
mark peer as block-aware after receiving message

dknopik committed:
merge from unstable

dknopik committed:
add remaining verification steps

dknopik committed:
add unit tests

dknopik committed:
fix places relying on header in sidecar

dknopik committed:
fix build

dknopik committed:
merge from unstable

dknopik commented:
Added some unit test with claude, let me know if it is a bit too much :)

dknopik committed:
update libp2p

dknopik committed:
merge from unstable

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,480 @@
+#[educe(PartialEq, Eq, Hash(bound = "E: EthSpec"))]
+pub struct PartialDataColumnSidecar<E: EthSpec> {
+    pub cells_present_bitmap: CellBitmap<E>,
+    pub column: VariableList<Cell<E>, <E as EthSpec>::MaxBlobCommitmentsPerBlock>,
```
nit:

```suggestion
    pub column: VariableList<Cell<E>, E::MaxBlobCommitmentsPerBlock>,
```

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,480 @@
+        .len()
+    }
+
+    pub fn max_size(max_blobs_per_block: usize) -> usize {
```
This is dead code, can we remove this?

If we keep this, I think we could use some documentation here - if im understanding this correctly, this type is essentially used as an union type - hence the max_size, is the full cell-only sidecar, right?

dknopik replied:
Removed - I believe I originally added this because I oriented myself along the code for full columns, but this ended up being unneeded.

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,480 @@
+}
+
+impl<E: EthSpec> PartialDataColumnSidecar<E> {
+    pub fn min_size() -> usize {
```
I can't find usage of this - is this required

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,480 @@
+        .len()
+    }
+
+    pub fn size(present_blobs: usize, block_blobs: usize) -> usize {
```
Same as above

dknopik committed:
merge from unstable

jimmygchen commented on beacon_node/beacon_chain/src/partial_data_column_assembler.rs:
```diff
@@ -0,0 +1,548 @@
+            };
+
+            // Check if merged column is now complete by trying to convert into full
+            let column = if let Some(full_column) = merged.try_clone_full(&header) {
```
Just a note that `try_clone_full` is a deep clone and it's called inside the `self.assemblies` write lock. If lock contention becomes an issue, this may be the place to optimise.

jimmygchen commented on beacon_node/beacon_chain/src/partial_data_column_assembler.rs:
```diff
@@ -0,0 +1,548 @@
+        &self,
+        block_root: Hash256,
+        header: &Arc<PartialDataColumnHeader<E>>,
+    ) -> Option<Vec<KzgVerifiedCustodyPartialDataColumn<E>>> {
```
I don't see where `None` is returned, do we need `Option` here?

dknopik replied:
Good catch, this was previously necessary - will directly return the Vec

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,480 @@
+                    // Keep this cell
+                    let cell = self.column.get(column_idx)?;
+                    new_column.push(cell);
+                    let proof = self.kzg_proofs.get(column_idx)?;
```
The early returns `?` here are problematic I think - that's a data integrity error and this should probably return an error and may need to penalise the peer for sending this? Otherwise, we're treating corrupted data the same as nothing passing the filter, and we may end up propagating these invalid messages.

dknopik replied:
Yep, agree.

dknopik replied:
Done in https://github.com/sigp/lighthouse/pull/8314/commits/a0ffc72e20c2d1d2f91664864f36b80fd04d9be9

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,480 @@
+                }
+                column_idx = column_idx
+                    .checked_add(1)
+                    .expect("Will not have more cells than 2^64 - 1");
```
Panic here isn't possible right now (I think, but i haven't finished), but ideally we return an error and log a crit instead - even if a bug is introduced in the future in parital messages, the beacon node can still gracefully fallback to receiving full sidecars.

jimmygchen replied:
Same applies to all the other `.expect`s in this file

dknopik replied:
Understandable. But if this is really triggered, we do not want a simple `UnexpectedOverflow` enum variant, because we really need to understand where exactly the overflow occurred.

I am actually thinking about adding `map` and `filter` functions  (or a combined  `filter_map`) to the ssz types, as encapsulating the fact that filtering values from a `VariableList` will not overflow it if capacity stays the same in there would allow far cleaner code in here.

jimmygchen replied:
Yeah, this is a good idea.

I agree we'd want some error logging if this happens, although there's a chance that the beacon node can recover and fallback to receiving full data columns?

As a temporary solution (if we don't get the SSZ changes in time) i think we should remove the `.expect`s. We don't want to set a precedent for their usage, and we intend to enforce this with lint.

dknopik replied:
Done in https://github.com/sigp/lighthouse/pull/8314/commits/a0ffc72e20c2d1d2f91664864f36b80fd04d9be9

jimmygchen reviewed (commented):
@dknopik I've started reviewing but haven't been able to get through this yet and I need to do some Gloas review now, so will come back to this shortly. 

I'll submit what I have so far, but I don't have the full picture yet, so some of my draft comments may not be correct

dknopik committed:
merge from unstable

dknopik committed:
address review

jimmygchen commented on Cargo.toml:
```diff
@@ -277,6 +279,7 @@ debug = true
-[patch."https://github.com/libp2p/rust-libp2p.git"]
-libp2p = { git = "https://github.com/sigp/rust-libp2p.git", rev = "defcaf1a78cf5b70a723b3fee0e0be051c1dbd88" }
-libp2p-mplex = { git = "https://github.com/sigp/rust-libp2p.git", rev = "defcaf1a78cf5b70a723b3fee0e0be051c1dbd88" }
+# This is intentionally unset as we use jxs fork in this PR. TODO(dknopik): restore/remove
```
Just adding a comment here as a reminder to update before we merge

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,446 @@
+        let mut new_bitmap = self.cells_present_bitmap.clone();
+        let mut new_column = Vec::new();
+        let mut new_proofs = Vec::new();
+        let mut column_idx = 0;
```
The name `column_idx` is a bit confusing here, worth adding a comment noting this is not the `ColumnIndex` of the data column sidecar, or perhaps renaming to something like `column_vec_idx` to differntiate.

dknopik replied:
Yeah, or `cell_idx`. Good point!

dknopik replied:
Done in https://github.com/sigp/lighthouse/pull/8314/commits/a0ffc72e20c2d1d2f91664864f36b80fd04d9be9

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,446 @@
+        })
+    }
+
+    pub fn merge(&self, other: &Self) -> Option<Self> {
```
Could use some doc here, not clear why this returns an `Option`

dknopik replied:
Partially done in https://github.com/sigp/lighthouse/commit/a0ffc72e20c2d1d2f91664864f36b80fd04d9be9, 
does not return an Option anymore. Will also add doc comment

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,446 @@
+    }
+
+    pub fn merge(&self, other: &Self) -> Option<Self> {
+        let new_bitmap = self.cells_present_bitmap.union(&other.cells_present_bitmap);
```
is it worth adding a length check here on both sidecars? i think it makes more sense if this function returns a `Result` - `Ok` if successfully merged; `Error` otherwise, e.g. header mismatch, length mismatch etc. This would also replace the `.expect()` with `?`.

I could be wrong on what's already been validated - either way i think some docs describing the assumptions would help.

dknopik replied:
Done in https://github.com/sigp/lighthouse/pull/8314/commits/a0ffc72e20c2d1d2f91664864f36b80fd04d9be9

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,446 @@
+                        .push(self.column.get(self_cell_idx)?.clone())
+                        .expect("Has same capacity");
+                    new_proofs
+                        .push(*self.kzg_proofs.get(self_cell_idx)?)
```
if we use the iterator's `.next()` for these lists, then we don't need to maintain the indices and would make it a bit more readable IMO.

dknopik replied:
Done in https://github.com/sigp/lighthouse/pull/8314/commits/a0ffc72e20c2d1d2f91664864f36b80fd04d9be9

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,446 @@
+                self.header.clone()
+            } else {
+                other.header.clone()
+            },
```
I guess an extra header equality check here is not required here, as the header is already validated during gossip verification?

dknopik replied:
Yeah. I guess validation is not enforced though, so it might make sense to move this to the `KzgVerifiedPartialDataColumn` instead?

jimmygchen replied:
yeah make sense to do the verification there

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,446 @@
+                .message()
+                .body()
+                .kzg_commitments_merkle_proof()?
+                .clone(),
```
Unnecessary clone i think?

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,446 @@
+            return None;
+        }
+
+        Some(DataColumnSidecar::Fulu(DataColumnSidecarFulu {
```
Reminder: will need to add gloas support

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -198,25 +202,80 @@ impl From<BeaconStateError> for GossipDataColumnError {
+    ///
+    /// ## Peer scoring
+    /// The column sidecar is invalid and the peer is faulty
+    InconsistentMaxCount {
```
This name is slightly confusing, what about
`InconsistentCommitmentsLength`?

dknopik replied:
Done in https://github.com/sigp/lighthouse/pull/8314/commits/00e9c83ee94a83789c7fcef551e55deb93ef956a

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -263,7 +322,9 @@ impl<T: BeaconChainTypes, O: ObservationStrategy> GossipVerifiedDataColumn<T, O>
         if chain
             .data_availability_checker
-            .is_data_column_cached(&column_sidecar.block_root(), &column_sidecar)
+            .determine_missing_cells_full(&column_sidecar.block_root(), column_sidecar.as_ref())
```
Why do we need to check the `parital_assembler`, wouldn't completed partials always be in the `availability_cache`? or is there a race?

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -475,10 +696,37 @@ pub fn verify_kzg_for_data_column<E: EthSpec>(
+        .sidecar
+        .filter(|idx| unverified_cells.contains(&idx))
+    else {
+        return Err(GossipDataColumnError::PriorKnownUnpublished.into());
```
I don't think this variant is applicable in a kzg verification function?

This error variant means: we have the blobs from the EL but we haven't publish them; if we get it via gossip, we should accept and forward them.

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -518,13 +766,15 @@ pub fn validate_data_column_sidecar_for_gossip_fulu<T: BeaconChainTypes, O: Obse
+    if let Some(cells) = chain
         .data_availability_checker
-        .is_data_column_cached(&data_column.block_root(), &data_column)
+        .determine_missing_cells_full(&data_column.block_root(), data_column.as_ref())
```
same as the other one - would the original `is_data_column_cached` be insufficient here?

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -562,6 +816,123 @@ pub fn validate_data_column_sidecar_for_gossip_fulu<T: BeaconChainTypes, O: Obse
+        // There is no header, so we check if we have a cached one to use
+        let Some(header) = chain
+            .data_availability_checker
+            .partial_assembler()
```
we could just `?` here instead of using `.and_then()` so it doesn't incorrectly return the `MissingHeader` error if `partial_assembler` is `None`.

dknopik replied:
Done something similar in https://github.com/sigp/lighthouse/pull/8314/commits/40125cdf58826d002670b2edfc3202e9110b90e7 - I introduced an error for missing partial assembler.

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -562,6 +816,123 @@ pub fn validate_data_column_sidecar_for_gossip_fulu<T: BeaconChainTypes, O: Obse
+            );
+        };
+
+        // If there was no header, there must be at least one cell.
```
I think you meant "if there is a header"?

dknopik replied:
Clarified in call, comment is correct

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -562,6 +816,123 @@ pub fn validate_data_column_sidecar_for_gossip_fulu<T: BeaconChainTypes, O: Obse
+        .data_availability_checker
+        .determine_missing_cells_partial(&block_root, &column)
+    else {
+        // TODO(dknopik): This shuld be a different error
```
Should be addressed before merging

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -985,4 +1369,240 @@ mod test {
+    }
+
+    #[tokio::test]
+    async fn partial_empty_message_without_cells_returns_error() {
```
All tests in `beacon_chain` run per-fork via the `FORK_NAME` env var in CI. I think we want to make tests either:
1. generic to forks via `test_spec()` so they run on every fork; OR:
2. just test one specific fork (e.g. `return` if `fork_name_from_env() != Some(Fulu)`)

jimmygchen replied:
minor note: `BeaconChainHarness` is pretty expensive to create, and it piles up pretty quickly if we use it to check all cases. If we need the `chain` to test the gossip checks, we could have them all in a single test as a  pragmatic solution, or do unit tests instead - but may need some refactor to make it testable.

dknopik replied:
both done in https://github.com/sigp/lighthouse/pull/8314/commits/edfa71c6d308fc6c008ed0ed95e147866c6af5c4

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,446 @@
+    pub cells_present_bitmap: CellBitmap<E>,
+    pub column: VariableList<Cell<E>, E::MaxBlobCommitmentsPerBlock>,
+    pub kzg_proofs: VariableList<KzgProof, E::MaxBlobCommitmentsPerBlock>,
+    pub header: VariableList<PartialDataColumnHeader<E>, U1>,
```
`ssz_types` 0.14.1 is out, I guess you wanted to use the new `ListEncodedOption` type here?
https://github.com/sigp/ssz_types/pull/83

dknopik replied:
Done in https://github.com/sigp/lighthouse/pull/8314/commits/ca4b5f0003e3d4cff44eeaa46725caf55b9b537c

dknopik committed:
merge from unstable

dknopik committed:
fix lint

dknopik committed:
track latest cell timestamp

dknopik committed:
fix getBlobs when partial messages are not enabled

dknopik committed:
update ssz_types

dknopik committed:
Arc/Box some more

This avoids some Clippy warnings in future commits and makes it more
consistent with full columns

dknopik committed:
Use ListEncodedOption

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,446 @@
+            match presence_bits {
+                (false, false) => {}
+                (true, other) => {
+                    new_column
```
Can we add a comment here explaining that both sides are KZG-verified so overlapping cells are guaranteed identical?

jimmygchen commented on beacon_node/lighthouse_network/src/types/partial.rs:
```diff
@@ -0,0 +1,498 @@
+                let send = self
+                    .partial_column
+                    .sidecar
+                    .filter(|idx| want.get(idx).expect("Bound checked above"))
```
This isn't reachable right now but we should avoid panics

jimmygchen commented on beacon_node/beacon_chain/src/partial_data_column_assembler.rs:
```diff
@@ -0,0 +1,548 @@
+        }
+    }
+
+    /// Returns true unless the header was already contained or the passed argument failed to convert to a header
```
> the passed argument failed to convert to a header

I don't see this? is this outdated comment

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -466,6 +633,60 @@ impl<E: EthSpec> KzgVerifiedCustodyDataColumn<E> {
+        self.data.index
+    }
+
+    pub fn merge(&self, other: &Self) -> Option<Self> {
```
Please add doc here

jimmygchen commented on beacon_node/network/src/network_beacon_processor/gossip_methods.rs:
```diff
@@ -193,6 +201,19 @@ impl<T: BeaconChainTypes> NetworkBeaconProcessor<T> {
         })
     }
 
+    /// Send a message on `message_tx` that  `peer_id` should be propagated on
```
This comment is incorrect i think?

dknopik committed:
better validation flow around determine_missing_cells

dknopik committed:
share harness across tests

dknopik committed:
Rename InconsistentMaxCount

dknopik committed:
Introduce error for missing column assembler

dknopik committed:
get rid of expects

dknopik committed:
merge from unstable

jimmygchen commented on beacon_node/lighthouse_network/src/types/partial.rs:
```diff
@@ -0,0 +1,502 @@
+        // For now, always request all cells
+        let mut request = partial_column.sidecar.cells_present_bitmap.clone();
+        for idx in 0..request.len() {
+            request
```
request seems unused after this?

dknopik replied:
oof.

good catch. 🤦

jimmygchen commented on beacon_node/lighthouse_network/src/types/partial.rs:
```diff
@@ -0,0 +1,502 @@
+        }
+        let metadata = PartialDataColumnPartsMetadata::<E> {
+            available: partial_column.sidecar.cells_present_bitmap.clone(),
+            request: partial_column.sidecar.cells_present_bitmap.clone(),
```
I don't understand this, why is available and request set to the same value?

jimmygchen commented on beacon_node/lighthouse_network/src/service/mod.rs:
```diff
@@ -886,6 +920,66 @@ impl<E: EthSpec> Network<E> {
+                // add to metrics
+                if let Some(v) = metrics::get_int_gauge(
+                    &metrics::FAILED_PUBLISHES_PER_MAIN_TOPIC,
+                    &[&format!("{:?}", topic.kind())],
```
might be worth tracking the partial publishing failures separately, perhaps extra metric so we know how many of the failures are partial / full if it occrus.

jimmygchen commented on beacon_node/lighthouse_network/src/types/pubsub.rs:
```diff
@@ -464,6 +464,35 @@ impl<E: EthSpec> PubsubMessage<E> {
 }
 
+/// Decodes incoming partial data column sidecar from gossipsub partial protocol.
+/// Note: Currenly, data columns are the only supported partial messages. In future this could
```
nit: typo
```suggestion
/// Note: Currently, data columns are the only supported partial messages. In future this could
```

jimmygchen commented on beacon_node/lighthouse_network/src/types/pubsub.rs:
```diff
@@ -464,6 +464,35 @@ impl<E: EthSpec> PubsubMessage<E> {
+                .map_err(|e| format!("Error decoding sidecar: {:?}", e))?;
+            let data_column = PartialDataColumn {
+                block_root,
+                // Partial messages are spec'd under the assumption that there is one column per subnet.
```
This is true for the current mainnet config, however PeerDAS has subnet decoupled from num of columns to allow network parameter tweaks, which means we supprot having having different values for:
- `DATA_COLUMN_SIDECAR_SUBNET_COUNT`
- `NUMBER_OF_COLUMNS`

So this isn't compatible with the partials spec.

dknopik replied:
I added a check on startup that these values are identical. Partial messages are not enabled otherwise.

jimmygchen commented on beacon_node/src/config.rs:
```diff
@@ -110,6 +110,11 @@ pub fn get_config<E: EthSpec>(
 
+    if parse_flag(cli_args, "enable-partial-columns") {
+        client_config.network.enable_partial_columns = true;
+        client_config.chain.enable_partial_columns = true;
```
Please add a cli test here https://github.com/sigp/lighthouse/blob/d7c78a7f89bd9f3a3330ef60d8a4e58959f73227/lighthouse/tests/beacon_node.rs#L2860

jimmygchen commented on beacon_node/lighthouse_network/src/service/mod.rs:
```diff
@@ -918,6 +1012,16 @@ impl<E: EthSpec> Network<E> {
+        propagation_source: PeerId,
+        topic: GossipTopic,
+    ) {
+        self.gossipsub_mut()
```
Be good to add the `GOSSIP_UNACCEPTED_MESSAGES_PER_CLIENT` metric here i think

jimmygchen commented on beacon_node/lighthouse_network/src/service/mod.rs:
```diff
@@ -1290,6 +1394,53 @@ impl<E: EthSpec> Network<E> {
+                            "Could not decode gossipsub partial message topic"
+                        );
+                    })
+                    .ok()?;
```
we shouldn't swallow decoding error here - should probably penalise peer

jimmygchen commented on beacon_node/lighthouse_network/src/types/pubsub.rs:
```diff
@@ -464,6 +464,35 @@ impl<E: EthSpec> PubsubMessage<E> {
+    group: &[u8],
+    data: &[u8],
+) -> Result<PartialDataColumn<E>, String> {
+    match topic.kind() {
```
not a concern for this PR but i'd imagine we need to decode with the `topic.fork_digest` to support Gloas

jimmygchen commented on beacon_node/lighthouse_network/Cargo.toml:
```diff
@@ -21,11 +21,12 @@ ethereum_ssz_derive = { workspace = true }
 itertools = { workspace = true }
 libp2p = { workspace = true }
-libp2p-mplex = { git = "https://github.com/libp2p/rust-libp2p.git" }
+libp2p-mplex = { git = "https://github.com/jxs/rust-libp2p.git", branch = "gossipsub-partial-messages" }
```
just a reminder comment to use canonical fork once merged

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,486 @@
+#[derive(Debug, Clone, Encode, Decode, PartialEq, Eq)]
+pub struct PartialDataColumnPartsMetadata<E: EthSpec> {
+    pub available: CellBitmap<E>,
+    pub request: CellBitmap<E>,
```
Doesnt' really impact SSZ encoding, but would be good to match the `requests` name in the spec .

jimmygchen commented on beacon_node/beacon_chain/src/metrics.rs:
```diff
@@ -1675,6 +1675,56 @@ pub static DATA_COLUMN_SIDECAR_GOSSIP_VERIFICATION_TIMES: LazyLock<Result<Histog
+    LazyLock::new(|| {
+        try_create_int_counter(
+            "beacon_partial_data_column_sidecar_header_processing_successes_total",
+            "Number of partial data column sidecars verified for gossip (excluding dupes)",
```
help description is same as above metric and needs update here

jimmygchen commented on beacon_node/beacon_chain/src/metrics.rs:
```diff
@@ -1675,6 +1675,56 @@ pub static DATA_COLUMN_SIDECAR_GOSSIP_VERIFICATION_TIMES: LazyLock<Result<Histog
+> = LazyLock::new(|| {
+    try_create_histogram(
+        "beacon_partial_data_column_sidecar_header_gossip_verification_seconds",
+        "Full runtime of partial data column sidecars gossip verification",
```
same description as `PARTIAL_DATA_COLUMN_SIDECAR_GOSSIP_VERIFICATION_TIMES`, perhaps add `header` ?

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,486 @@
+        })
+    }
+
+    fn len(&self) -> Result<usize, PartialDataColumnSidecarError> {
```
perhaps rename this to `verify_len`?

jimmygchen commented on beacon_node/network/src/sync/manager.rs:
```diff
@@ -141,6 +141,9 @@ pub enum SyncMessage<E: EthSpec> {
     UnknownParentDataColumn(PeerId, Arc<DataColumnSidecar<E>>),
 
+    /// A partial data column with an unknown parent has been received.
+    UnknownParentPartialDataColumn(PeerId, Hash256, Hash256, Slot),
```
be worth documenting the two roots fields here for clarity

dknopik replied:
converted into struct variant instead

jimmygchen commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -3124,6 +3187,95 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+        let sampling_columns = self
+            .data_availability_checker
+            .custody_context()
+            .sampling_columns_for_epoch(slot.epoch(T::EthSpec::slots_per_epoch()), &self.spec);
```
nit: could just use `self.sampling_columns_for_epoch`

jimmygchen commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -338,6 +431,12 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
+        if let Some(assembler) = &self.partial_assembler
+            && let Ok(header) = block.try_into()
+        {
+            assembler.init(block.canonical_root(), Arc::new(header));
```
Can use block root here without recomputing `executed_block.import_data.block_root`

jimmygchen commented on beacon_node/network/src/sync/manager.rs:
```diff
@@ -903,6 +906,21 @@ impl<T: BeaconChainTypes> SyncManager<T> {
+                    block_root,
+                    parent_root,
+                    slot,
+                    BlockComponent::PartialDataColumn(DownloadResult {
```
Do we need a new variant for partials? we handle them exactly the same way, and could just use `DataColumns`, but it does carry extra information if we do log this enum somewhere.

jimmygchen replied:
it probably doesn't matter - historically we had the data column in the enum variant and we store them, but we no longer do that.

jimmygchen commented on beacon_node/network/src/network_beacon_processor/sync_methods.rs:
```diff
@@ -218,9 +218,15 @@ impl<T: BeaconChainTypes> NetworkBeaconProcessor<T> {
                 // to be sent from the peers if we already have them.
-                let publish_blobs = false;
-                self.fetch_engine_blobs_and_publish(signed_beacon_block, block_root, publish_blobs)
+                if let Ok(header) = signed_beacon_block.as_ref().try_into() {
```
just a note that the error path is not reachable today because `try_into` only returns error if the block is pre-deneb. I'd usually avoid ignoring error but in this case its probably fine.

jimmygchen commented on beacon_node/network/src/network_beacon_processor/mod.rs:
```diff
@@ -251,6 +251,32 @@ impl<T: BeaconChainTypes> NetworkBeaconProcessor<T> {
+        let processor = self.clone();
+        let process_fn = async move {
+            processor
+                .process_gossip_partial_data_column_sidecar(
```
We discussed about this yesterday - but worth noting here that we are not passing subnet id to processing because the partial message design assumes column index == subnet id. This is true for mainnet config today, but isn't true for the peerdas design as it supports subnet coupling and these network parameters can be tweaked via configuration. 

we should add the number of subnet & number of column validation check -  probably to where we parse the cli flag?

dknopik committed:
Metadata requests field matches spec name

dknopik committed:
fix horrendous bug

dknopik committed:
add cli test

dknopik committed:
reuse block root

dknopik committed:
check for 1 to 1 mapping of subnet and column

dknopik committed:
use self.sampling_columns_for_epoch

dknopik committed:
merge from unstable

jimmygchen commented:
I'm pretty close to finishing the review. 

I got claude to run a round of preliminary tests and looked at metrics, logs and traces on a local devnet - here's the report: https://hackmd.io/@jimmygchen/BJqjwMCnbx

TL;DR this seems to be working fine with unstable on a local network.

dknopik committed:
struct variant for UnknownParentPartialDataColumn

dknopik committed:
rename len into verify_len

dknopik committed:
fix metric definitions

dknopik committed:
penalize on undecodable topic

dknopik committed:
add GOSSIP_UNACCEPTED_MESSAGES_PER_CLIENT

dknopik committed:
nit typo

dknopik committed:
fix comment on init

dknopik committed:
fix comment on propagate_partial_validation_failure

dknopik committed:
FAILED_PARTIAL_PUBLISHES_PER_MAIN_TOPIC

dknopik committed:
move merge to wrapper

dknopik committed:
merge from unstable

dknopik committed:
fix copy paste

dknopik committed:
merge from unstable

dknopik committed:
organize imports

jimmygchen commented on beacon_node/http_api/src/publish_blocks.rs:
```diff
@@ -514,15 +516,51 @@ fn publish_column_sidecars<T: BeaconChainTypes>(
-        .map_err(|_| BlockError::BeaconChainError(Box::new(BeaconChainError::UnableToPublish)))
+    let mut full_messages = Vec::new();
+    let mut partial_columns = Vec::new();
+    let mut partial_header = None;
```
We should get the partial header creation and publishing here if it's not enabled.

dknopik committed:
update libp2p and move to master branch

dknopik committed:
merge from unstable

dknopik committed:
fmt

dknopik committed:
fmt

dknopik committed:
avoid creating sidecar and header on publish if disabled

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -549,30 +923,45 @@ pub fn validate_data_column_sidecar_for_gossip_fulu<T: BeaconChainTypes, O: Obse
+            MissingCellsError::MismatchesCachedColumn => {
+                GossipDataColumnError::MismatchesCachedColumn
+            }
+            MissingCellsError::UnexpectedError(_) => todo!("handle unexpected error"),
```
This is in production code path although unreachable unless partial flag is enabled. We should just return an error here.

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -595,6 +984,137 @@ pub fn validate_data_column_sidecar_for_gossip_fulu<T: BeaconChainTypes, O: Obse
+                header,
+            };
+        }
+        Err(MissingCellsError::UnexpectedError(e)) => todo!("handle unexpected error {:?}", e),
```
same as above

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -262,22 +340,29 @@ impl<T: BeaconChainTypes, O: ObservationStrategy> GossipVerifiedDataColumn<T, O>
+                Err(GossipDataColumnError::MismatchesCachedColumn)
+            }
+            Err(MissingCellsError::UnexpectedError(_)) => {
+                todo!("handle unexpected error")
```
return error

jimmygchen commented on consensus/types/src/data/partial_data_column_sidecar.rs:
```diff
@@ -0,0 +1,385 @@
+
+    pub fn get(&self, idx: usize) -> Option<(&Cell<E>, &KzgProof)> {
+        self.column
+            .get(idx)
```
`idx` is blob index, but `self.column` may not have all cells so the index may not match?

jimmygchen replied:
https://github.com/sigp/lighthouse/pull/8314/commits/0e075c542c46eb34c7bd018aa98454c4ef034941

jimmygchen reviewed (commented):
Hi @dknopik this is looking pretty good now, I reckon we merge this and address any remaining issues separately - I'm going to do a quick final pass and fix CI issues and merge.

I think this comment is needs a fix if this is true:
> idx is blob index, but self.column may not have all cells so the index may not match?

jimmygchen committed:
Fix fmt

jimmygchen committed:
Merge remote-tracking branch 'origin/unstable' into fork/dknopik/partial-columns

jimmygchen commented on beacon_node/beacon_chain/src/data_column_verification.rs:
```diff
@@ -262,22 +340,29 @@ impl<T: BeaconChainTypes, O: ObservationStrategy> GossipVerifiedDataColumn<T, O>
+        match chain
             .data_availability_checker
-            .is_data_column_cached(&column_sidecar.block_root(), &column_sidecar)
+            .missing_cells_for_column_sidecar(&column_sidecar)
```
I might be wrong but i think we should just treat the data column as not cached and proceed with publishing the column anyway.

I think this is probably worth iterating further - but for now, before partials is widely deployed on the network, we should prioritise having full columns published, even when there's an error in partials.

jimmygchen replied:
I couldn't do the error handling cleanly without changing the function. Let's discuss this, but it shouldn't block the PR as it's partials only path.

jimmygchen committed:
Fix fmt

jimmygchen committed:
Fix `PartialDataColumnSidecar::get()` blob index mapping.

jimmygchen reviewed (approved):
@dknopik thanks for your patience with this! 

~~I'm happy to merge this once this is updated to use the canonical fork:~~ (**EDIT:** The below has been addressed, I have delete the outdated TODO comment)
  - [`Cargo.toml:279`](https://github.com/sigp/lighthouse/pull/8314#discussion_r3019599356) - Reminder to update dependency before merge

Happy to merge this as is, and the remaining comments below to be addressed in a follow up PR:
* `todo!()`/`expect() in partial-enabled code paths:
  * [`data_column_verification.rs:938`](https://github.com/sigp/lighthouse/pull/8314#discussion_r3125304697)   
  * [`data_column_verification.rs:1079`](https://github.com/sigp/lighthouse/pull/8314#discussion_r3125305939)
  * [`data_column_verification.rs:363`](https://github.com/sigp/lighthouse/pull/8314#discussion_r3125307522) 
  * [`partial.rs:213`](https://github.com/sigp/lighthouse/pull/8314#discussion_r3071336258)
* `missing_cells_for_column_sidecar` handling
  * [`data_column_verification.rs:345`](https://github.com/sigp/lighthouse/pull/8314#discussion_r3131832010)

jimmygchen committed:
Fix syn duplicate in Cargo.lock (see 9f3873f2bf) and remove comment code.
